File: src/api/tauri/process.ts
/**
 * Typed Tauri IPC wrappers for Process Intelligence commands.
 *
 * These are the ONLY entry points for Tauri `invoke` calls related to process
 * listing and control. Components and pages MUST NOT call `invoke` directly.
 *
 * Command names are snake_case (Tauri convention).
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  ProcessDeepDetail,
  ProcessInfo,
  ProcessKillResult,
  ProcessSnapshot,
  ProcessTreeNode,
  ServiceInfo,
} from '../../types/device.types';

/** List running processes with risk classification, highest impact first. */
export const listProcesses = (topN?: number): Promise<ProcessSnapshot> =>
  invoke<ProcessSnapshot>('list_processes', { topN: topN ?? null });

/** Process tree forest for explorer tree views. */
export const getProcessTree = (max?: number): Promise<ProcessTreeNode[]> =>
  invoke<ProcessTreeNode[]>('get_process_tree', { max: max ?? null });

/** Detail for a single process id, or null when not found. */
export const getProcessDetail = (pid: number): Promise<ProcessInfo | null> =>
  invoke<ProcessInfo | null>('get_process_detail', { pid });

/**
 * Deep process detail: memory regions, wait chains, token, named handles.
 * Full results may require running elevated.
 */
export const getProcessDeep = (pid: number): Promise<ProcessDeepDetail> =>
  invoke<ProcessDeepDetail>('get_process_deep', { pid });

/**
 * Terminate a process. Requires `confirm: true`.
 * Optional `tree` ends the process and its descendants.
 * Protected system processes may be refused by the backend.
 */
export const killProcess = (
  pid: number,
  confirm: boolean,
  tree?: boolean,
): Promise<ProcessKillResult> =>
  invoke<ProcessKillResult>('kill_process', {
    pid,
    confirm,
    tree: tree ?? false,
  });

/** Windows services inventory for the Process Explorer services tab. */
export const listServices = (): Promise<ServiceInfo[]> =>
  invoke<ServiceInfo[]>('list_services');

File: src-tauri/src/commands/process.rs
//! Process Intelligence IPC (Process Hacker / Sysinternals direction).

use tauri::State;

use crate::error::CoreError;
use crate::models::{
    ProcessDeepDetail, ProcessInfo, ProcessKillResult, ProcessSnapshot, ProcessTreeNode,
    ServiceInfo,
};
use crate::process;
use crate::AppState;

/// Lists live processes sorted by memory then CPU. Optional `top_n` (default 80).
#[tauri::command]
pub fn list_processes(
    _state: State<'_, AppState>,
    top_n: Option<usize>,
) -> Result<ProcessSnapshot, CoreError> {
    process::list_processes(top_n)
}

/// Process tree forest for explorer views.
#[tauri::command]
pub fn get_process_tree(
    _state: State<'_, AppState>,
    max: Option<usize>,
) -> Result<Vec<ProcessTreeNode>, CoreError> {
    process::process_tree(max)
}

/// Detail for a single process id (includes modules/threads/handles when available).
#[tauri::command]
pub fn get_process_detail(
    _state: State<'_, AppState>,
    pid: u32,
) -> Result<Option<ProcessInfo>, CoreError> {
    process::get_process(pid)
}

/// Terminates a process. Requires `confirm: true`. Optional process-tree kill.
#[tauri::command]
pub fn kill_process(
    state: State<'_, AppState>,
    pid: u32,
    confirm: bool,
    tree: Option<bool>,
) -> Result<ProcessKillResult, CoreError> {
    let conn = state.conn()?;
    process::kill_process(Some(&conn), pid, confirm, tree.unwrap_or(false))
}

/// Windows services inventory.
#[tauri::command]
pub fn list_services(_state: State<'_, AppState>) -> Result<Vec<ServiceInfo>, CoreError> {
    process::list_services()
}

/// Deep process detail: memory map, wait chains, token privileges, named handles.
#[tauri::command]
pub fn get_process_deep(
    _state: State<'_, AppState>,
    pid: u32,
) -> Result<ProcessDeepDetail, CoreError> {
    process::deep::get_process_deep(pid)
}
